app/uk/gov/hmrc/transitmovements/controllers/MovementsController.scala

```diff
@@ -82,38 +88,97 @@ class MovementsController @Inject() (
-        message <- messageFactory
-          .create(MessageType.ArrivalNotification, arrivalData.generationDate, received, None, request.body, MessageStatus.Processing)
-          .asPresentation
-        movement = movementFactory.createArrival(eori, MovementType.Arrival, arrivalData, message, received, received)
-        _ <- repo.insert(movement).asPresentation
-      } yield MovementResponse(movement._id, Some(movement.messages.head.id))).fold[Result](
-        baseError => Status(baseError.code.statusCode)(Json.toJson(baseError)),
-        response => Ok(Json.toJson(response))
-      )
+  private def createArrival(eori: EORINumber): Action[Source[ByteString, _]] = Action.streamWithSize {
+    (request: Request[Source[ByteString, _]], size: Long) =>
+      {
+        implicit val hc: HeaderCarrier = HeaderCarrierConverter.fromRequest(request)
+        if (smallMessageLimitService.checkContentSize(size)) createArrivalWithObjectStore(eori, request)
+        else {
+          for {
+            arrivalData <- movementsXmlParsingService.extractArrivalData(request.body).asPresentation
+            received   = OffsetDateTime.ofInstant(clock.instant, ZoneOffset.UTC)
+            movementId = MovementId(ShortUUID.next(clock, random))
+            message <- messageFactory
+              .create(MessageType.ArrivalNotification, arrivalData.generationDate, received, None, request.body, MessageStatus.Processing)
+              .asPresentation
+            movement = movementFactory.createArrival(movementId, eori, MovementType.Arrival, arrivalData, message, received, received)
+            _ <- repo.insert(movement).asPresentation
+          } yield MovementResponse(movement._id, Some(movement.messages.head.id))
+        }
+      }.fold[Result](baseError => Status(baseError.code.statusCode)(Json.toJson(baseError)), response => Ok(Json.toJson(response)))
   }
 
-  private def createDeparture(eori: EORINumber): Action[Source[ByteString, _]] = Action.stream {
-    implicit request =>
-      (for {
-        declarationData <- movementsXmlParsingService.extractDeclarationData(request.body).asPresentation
-        received = OffsetDateTime.ofInstant(clock.instant, ZoneOffset.UTC)
-        message <- messageFactory
-          .create(MessageType.DeclarationData, declarationData.generationDate, received, None, request.body, MessageStatus.Processing)
-          .asPresentation
-        movement = movementFactory.createDeparture(eori, MovementType.Departure, declarationData, message, received, received)
-        _ <- repo.insert(movement).asPresentation
-      } yield MovementResponse(movement._id, Some(movement.messages.head.id))).fold[Result](
+  private def createArrivalWithObjectStore(
+    eori: EORINumber,
+    request: Request[Source[ByteString, _]]
+  )(implicit hc: HeaderCarrier): EitherT[Future, PresentationError, MovementResponse] =
+    for {
+      arrivalData <- movementsXmlParsingService.extractArrivalData(request.body).asPresentation
```

Comment:
insert in the repo first, use the returned object to then pass the ids to object store service

Reply:
We already had a discussion and decided to go with this approach.

Reply:
Cool, just for my understanding  - what was the logic behind the decision? Because if it doesn't save to object-store then we won't be saving it in the DB?

Reply:
To be fair, I did say "if that's the best way" and I'm not as close to the code as I once was -- I don't want to be totally prescriptive in how we do this.

In the context of small messages -- it is a bit chicken and egg. If we put it in the DB first then fail the call to object store, then we just need to ensure the status gets set to failed in the DB. If we put it in object store first, we need to pull id generation out.

From the point of view of retention, if we put it in OS then it'll hang around for 7 years, if we start with the DB that's 31 days. That might be a factor in decisions.

Finally, ideally, the controller wouldn't generate the IDs but I understand why in this case. I mentioned that reservation at the time we discussed it.

So, either approach is fine, but we do need to agree on it. Note that I do need to do some work in transit-movements around object store for platops anyway, so worst comes to worst we can tweak it later.

I do want to do it right, but remember we're agile too!

Reply:
If we put it in object-store first, then we would need to not fail the for comprehension if the call to object-store fails. Right now it won't create a movement in the DB if the call fails so just change `objectSummary <- objectStoreService.addMessage(movementId, messageId, request.body).asPresentation` to `objectSummary = objectStoreService.addMessage(movementId, messageId, request.body).asPresentation`

Reply:
For the small message route, that's okay though, because this will be from a sync call -- so if there is no DB entry, it doesn't matter because the trader should already know.

Reply:
Agree